include/dxc/HLSL/DxilSignatureAllocator.inl

```diff
@@ -332,117 +348,152 @@ unsigned DxilSignatureAllocator::PackOptimized(std::vector<PackElement*> element
-    clipcullTempElements[row].kind = clipcullElementsByRow[row][0]->GetKind();
-    clipcullTempElements[row].interpolation = clipcullElementsByRow[row][0]->GetInterpolationMode();
-    clipcullTempElements[row].interpretation = clipcullElementsByRow[row][0]->GetInterpretation();
-    clipcullTempElements[row].dataBitWidth = clipcullElementsByRow[row][0]->GetDataBitWidth();
-    clipcullTempElements[row].rows = 1;
-    clipcullTempElements[row].cols = clipcullComponentsByRow[row];
-  }
-
   // ==========
   // Allocate 4-component elements
   if (!vec4Elements.empty()) {
     std::sort(vec4Elements.begin(), vec4Elements.end(), CmpElementsLess);
-    unsigned used = PackGreedy(vec4Elements, startRow, numRows);
-    startRow += used;
-    numRows -= used;
-    if (rowsUsed < used)
-      rowsUsed = used;
+    rowsUsed = std::max(rowsUsed, PackGreedy(vec4Elements, startRow, numRows));
+    startRow = std::max(startRow, rowsUsed);
   }
 
   // ==========
   // Allocate indexed tessfactors in rightmost column
   if (!indexedtessElements.empty()) {
     std::sort(indexedtessElements.begin(), indexedtessElements.end(), CmpElementsLess);
-    unsigned used = PackGreedy(indexedtessElements, startRow, numRows, 3);
-    if (rowsUsed < used)
-      rowsUsed = used;
+    rowsUsed = std::max(rowsUsed, PackGreedy(indexedtessElements, startRow, numRows, 3));
   }
 
   // ==========
   // Allocate arbitrary
   if (!arbElements.empty()) {
     std::sort(arbElements.begin(), arbElements.end(), CmpElementsLess);
-    unsigned used = PackGreedy(arbElements, startRow, numRows);
-    if (rowsUsed < used)
-      rowsUsed = used;
+    rowsUsed = std::max(rowsUsed, PackGreedy(arbElements, startRow, numRows));
   }
 
   // ==========
   // Allocate system values
   if (!svElements.empty()) {
     std::sort(svElements.begin(), svElements.end(), CmpElementsLess);
-    unsigned used = PackGreedy(svElements, startRow, numRows);
-    if (rowsUsed < used)
-      rowsUsed = used;
+    rowsUsed = std::max(rowsUsed, PackGreedy(svElements, startRow, numRows));
   }
 
   // ==========
   // Allocate clip/cull
-  for (unsigned i = 0; i < clipcullRegUsed; ++i) {
-    bool bAllocated = false;
-    unsigned cols = clipcullComponentsByRow[i];
-    for (unsigned row = startRow; row < startRow + numRows; ++row) {
-      if (DetectRowConflict(&clipcullTempElements[i], row))
+  std::sort(clipcullElements.begin(), clipcullElements.end(), CmpElementsLess);
+  unsigned numClipCullComponents = 0;
+  unsigned clupCullMultiRowCols = 0;
+  for (auto &SE : clipcullElements) {
+    numClipCullComponents += SE->GetRows() * SE->GetCols();
+    if (SE->GetRows() > 1) {
+      clupCullMultiRowCols += SE->GetCols();
+    }
+  }
+  if (0 == clupCullMultiRowCols) {
+    // Preallocate clip/cull elements into two rows and allocate independently
+    DxilSignatureAllocator clipcullAllocator(2, m_bUseMinPrecision);
+    unsigned clipcullRegUsed = clipcullAllocator.PackGreedy(clipcullElements, 0, 2);
```

Comment:
It would be nice to have a constant for all of those twos. `MaxNumClipCullRows` or something.

Reply:
It turns out this already exists as `DXIL::kMaxClipOrCullDistanceElementCount`.  I've added this, but code would have to change if this wasn't == 2, so I've also added a static_assert.